### ISSUE TITLE
[Kiali 1.59.1] Error fetching Istiod resource threholds on Web UI

### ISSUE THREAD

**Issue by rondy-cg** [kiali/kiali#5742]
We just upgraded Istio from 1.13.2 to 1.16.0 on AWS EKS(v1.23) cluster, and Kiali was upgraded to 1.59.1 also.

When I port-forward Kiali to local (MacOS Ventura 13.1), and access it from Chrome browser,
we got several warnings and one error `Error fetching Istiod resource threholds` as below,
 does anyone see the similar issue? And any clue to resolve it?
![image](https://user-images.githubusercontent.com/102935179/213624065-00290827-512d-4b5e-8825-d5ac64fff335.png)

**Comment by jshaughn** [comment]
It's possible that the warnings are valid, given the changes in versions, so I'd suggest investigating those by looking at the Istio Config list and diving down into any warnings or errors.

The failure to fetch istiod information could be due to the istio debug endpoints not being exposed, maybe check the Kiali logs for more information.  New versions of Kiali require the ability to access the istiod API to be able to inspect the control plane.  (Note that in the upcoming Kiali v1.63 release we add explicit support to configure Kiali to not access the istio API, when users explicitly don't want it exposed.  This disables some Kiali features that depend on the API).

**Comment by rondy-cg** [comment]
@jshaughn 
Thanks for comments.
We doubted that the suspicious error might be related to the implementation below.
https://github.com/kiali/kiali/blob/v1.61.0/business/mesh.go#L638

```
2023/01/20 04:27:51 http: panic serving 127.0.0.1:36222: runtime error: invalid memory address or nil pointer dereference
goroutine 8212 [running]:
net/http.(*conn).serve.func1()
        /opt/hostedtoolcache/go/1.17.10/x64/src/net/http/server.go:1802 +0xb9
panic({0x1b0f040, 0x3069540})
        /opt/hostedtoolcache/go/1.17.10/x64/src/runtime/panic.go:1047 +0x266
github.com/kiali/kiali/business.(*MeshService).IstiodResourceThresholds(0xc002de40d8)
        /home/runner/work/kiali/kiali/business/mesh.go:638 +0x119
```

For further investigation, do you think if it's due to below or other thoughts?

1. No default setting for resource limit of istio deployment 
2. Couldn't discover the istio deployment or container due to name changes

Actually, I only installed istiod-1-16-0 as below and didn't change any setting on purpose, shall I change the resource limit?
```
% helm install istiod-1-16-0 istio/istiod \
    --set revision=1-16-0 \
    -n istio-system \
```

**Comment by rondy-cg** [comment]
Moreover, share some istioctl analyze errors which looks not critical though ... 

```
% istioctl analyze -A | grep -i Error
Error: Analyzers found issues when analyzing all namespaces.
See https://istio.io/v1.16/docs/reference/config/analysis for more information about causes and resolutions.
Error [IST0101] (Gateway default/hello-gateway) Referenced selector not found: "istio=hello-ingressgateway"
Error [IST0101] (Gateway istio-system/ingressgateway) Referenced selector not found: "istio=ingressgateway"
Error [IST0101] (Gateway istio-system/dummy-dps-trading-gateway) Referenced selector not found: "istio=dummy-dps-trading-ingressgateway"
Error [IST0101] (Gateway istio-system/dummy-gateway) Referenced selector not found: "istio=dummy-ingressgateway"
Error [IST0101] (Gateway istio-system/dummy-watcher-gateway) Referenced selector not found: "istio=dummy-watcher-ingressgateway"
Error [IST0106] (Gateway istio-system/dummy-dps-trading-gateway) Schema validation error: port names in servers must be unique: duplicate name tcp
```

Especially the last error "port names in servers must be unique" which was not there in istio 1.13.2, something new in istio 1.16.0?

**Comment by fl64** [comment]
I have the same problem with kiali, because due to certain circumstances I cannot set limits for istiod

**Comment by jshaughn** [comment]
cc @leandroberetta

**Comment by leandroberetta** [comment]
Hello,

> 
> For further investigation, do you think if it's due to below or other thoughts?
> 
> 1. No default setting for resource limit of istio deployment

You should be fine without limits, in that case the graph won't show threhold lines.

> 2. Couldn't discover the istio deployment or container due to name changes
> 
> Actually, we only installed istiod-1-16-0 as below and didn't change any setting on purpose, shall I change the resource limit?
> 
> ```
> % helm install istiod-1-16-0 istio/istiod \
>     --set revision=1-16-0 \
>     -n istio-system \
> ```

This has to be due to the name change, looking at the code, we are discovering the istiod deployment with a property configured in [Kiali CR](https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.external_services.istio.istiod_deployment_name). 

The way you upgraded Istio was with a revision, so after the installation and testing the revision became the stable, and I guess at least the property should be updated (the name of the deployment in the CR). Can you try updating the name?. 

Thanks for sharing this case with us, it really help us to understand if we have to add some documentation or code fixes for canary upgrades of the control plane.

**Comment by rondy-cg** [comment]
@leandroberetta @jshaughn 
Just FYI, not sure how it works, the following workarounds (additional changes on kiali.yaml) fixed the all the Kiali errors we were facing ...

```
       istio:
         root_namespace: istio-system
+        config_map_name: istio-1-16-0
+        istiod_deployment_name: istiod-1-16-0
```

**Comment by jshaughn** [comment]
Good news, thanks for letting us know the resolution.  @leandroberetta , that seems to make sense, right?  After the update the names are different.

**Comment by leandroberetta** [comment]
Yes, makes sense, and it's good to see the problems go away. I'm thinking that this needs to be somewhere in the docs. I will take a look to see where we can add it.

**Comment by fl64** [comment]
This solution doesn't help me, because Kiali tries to get limits, but they are absent, and panic again.

**Comment by fl64** [comment]
Hello!
No, for each istio version we use revision
```
kiali-7457486ccd-5kczt 2023/01/31 10:47:13 http: panic serving 10.111.0.220:43218: runtime error: invalid memory address or nil pointer dereference                                                                
kiali-7457486ccd-5kczt goroutine 201506 [running]:                                                                                                                                                                 
kiali-7457486ccd-5kczt net/http.(*conn).serve.func1()                                                                                                                                                              
kiali-7457486ccd-5kczt     /usr/local/go/src/net/http/server.go:1825 +0xbf                                                                                                                                         
kiali-7457486ccd-5kczt panic({0x1aadb40, 0x2ebadb0})                                                                                                                                                               
kiali-7457486ccd-5kczt     /usr/local/go/src/runtime/panic.go:844 +0x258                                                                                                                                           
kiali-7457486ccd-5kczt github.com/kiali/kiali/business.(*MeshService).IstiodResourceThresholds(0xc000d03098)                                                                                                       
kiali-7457486ccd-5kczt     /src/business/mesh.go:638 +0x119                                                                                                                                                        
kiali-7457486ccd-5kczt github.com/kiali/kiali/handlers.IstiodResourceThresholds({0x2060ef8, 0xc0013d4738}, 0xc000894690?)                                                                                          
kiali-7457486ccd-5kczt     /src/handlers/mesh.go:43 +0x94                                                                                                                                                          
kiali-7457486ccd-5kczt net/http.HandlerFunc.ServeHTTP(0x1d90705?, {0x2060ef8?, 0xc0013d4738?}, 0x40d2e5?)                                                                                                          
kiali-7457486ccd-5kczt     /usr/local/go/src/net/http/server.go:2084 +0x2f                                                                                                                                         
kiali-7457486ccd-5kczt github.com/kiali/kiali/routing.metricHandler.func1({0x206e758?, 0xc0002c8e00}, 0xc00052a301?)                                                                                               
kiali-7457486ccd-5kczt     /src/routing/router.go:145 +0x13b                                                                                                                                                       
kiali-7457486ccd-5kczt net/http.HandlerFunc.ServeHTTP(0x206fad0?, {0x206e758?, 0xc0002c8e00?}, 0xc00052a3e0?)                                                                                                      
kiali-7457486ccd-5kczt     /usr/local/go/src/net/http/server.go:2084 +0x2f                                                                                                                                         
kiali-7457486ccd-5kczt github.com/kiali/kiali/handlers.AuthenticationHandler.Handle.func1({0x206e758, 0xc0002c8e00}, 0xc00195ad00)                                                                                 
kiali-7457486ccd-5kczt     /src/handlers/authentication.go:78 +0x53e                                                                                                                                               
kiali-7457486ccd-5kczt net/http.HandlerFunc.ServeHTTP(0x19bc100?, {0x206e758?, 0xc0002c8e00?}, 0xc0000ff900?)                                                                                                      
kiali-7457486ccd-5kczt     /usr/local/go/src/net/http/server.go:2084 +0x2f                                                                                                                                         
kiali-7457486ccd-5kczt github.com/kiali/kiali/server.plainHttpMiddleware.func1({0x206e758?, 0xc0002c8e00?}, 0xc000894630?)                                                                                         
kiali-7457486ccd-5kczt     /src/server/server.go:152 +0x65                                                                                                                                                         
kiali-7457486ccd-5kczt net/http.HandlerFunc.ServeHTTP(0xc00195ac00?, {0x206e758?, 0xc0002c8e00?}, 0x0?)                                                                                                            
kiali-7457486ccd-5kczt     /usr/local/go/src/net/http/server.go:2084 +0x2f                                                                                                                                         
kiali-7457486ccd-5kczt github.com/gorilla/mux.(*Router).ServeHTTP(0xc0005c0900, {0x206e758, 0xc0002c8e00}, 0xc00195ab00)                                                                                           
kiali-7457486ccd-5kczt     /go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210 +0x1cf                                                                                                                             
kiali-7457486ccd-5kczt github.com/NYTimes/gziphandler.GzipHandlerWithOpts.func1.1({0x206e608, 0xc000b3a0e0}, 0x77839156ff?)                                                                                        
kiali-7457486ccd-5kczt     /go/pkg/mod/github.com/!n!y!times/gziphandler@v1.1.1/gzip.go:336 +0x23c                                                                                                                 
kiali-7457486ccd-5kczt net/http.HandlerFunc.ServeHTTP(0x7f97fab598d8?, {0x206e608?, 0xc000b3a0e0?}, 0x40d2e5?)                                                                                                     
kiali-7457486ccd-5kczt     /usr/local/go/src/net/http/server.go:2084 +0x2f                                                                                                                                         
kiali-7457486ccd-5kczt net/http.(*ServeMux).ServeHTTP(0x0?, {0x206e608, 0xc000b3a0e0}, 0xc00195ab00)                                                                                                               
kiali-7457486ccd-5kczt     /usr/local/go/src/net/http/server.go:2462 +0x149                                                                                                                                        
kiali-7457486ccd-5kczt net/http.serverHandler.ServeHTTP({0xc000894480?}, {0x206e608, 0xc000b3a0e0}, 0xc00195ab00)                                                                                                  
kiali-7457486ccd-5kczt     /usr/local/go/src/net/http/server.go:2916 +0x43b                                                                                                                                        
kiali-7457486ccd-5kczt net/http.(*conn).serve(0xc001376e60, {0x206fad0, 0xc00128f710})                                                                                                                             
kiali-7457486ccd-5kczt     /usr/local/go/src/net/http/server.go:1966 +0x5d7                                                                                                                                        
kiali-7457486ccd-5kczt created by net/http.(*Server).Serve                                                                                                                                                         
kiali-7457486ccd-5kczt     /usr/local/go/src/net/http/server.go:3071 +0x4db   
```
kiali 1.62, istio 1.16.2

**Comment by jshaughn** [comment]
@fl64, Is your istio API (the debug/ endpoints) blocked from being used by Kiali?  This may contribute to the problem you see.  Note that in v1.63 the Kiali CR now offers `istio_api_enabled`, and when set to false Kiali will not try to access the istiod api.  It does cause Kiali to drop some features (see https://kiali.io/docs/configuration/no-istiod/),

But, @leandroberetta, from the stack trace it seems that the following lines make a false assumption about teh settings you can expect when looking at the istioDeployment model:  https://github.com/kiali/kiali/blame/master/business/mesh.go#L650-L651

Several of those fields are marked as optional (and have omitempty in the json conversion).  I think some protection needs to be added. Do you not agree?

Finally, @josunect I think it is OK to still call `fetchIstiodResourceThresholds()` from the Overview page, even when `istio_api_enabled=false`, correct?  I don't think this uses the API, ity is just querying about the istiod pod itself.  But I wanted to check with you.

**Comment by leandroberetta** [comment]
> Several of those fields are marked as optional (and have omitempty in the json conversion). I think some protection needs to be added. Do you not agree?
> 

I agree 100%, I will add them, but yet I'm not sure why I usually run Istio without limits and works fine. 

Regarding the thresholds and the API, I confirmed is not using Istio API.

**Comment by fl64** [comment]
Sorry I'm late, missed this message. Here's the config
```
auth:
  strategy: anonymous
deployment:
  accessible_namespaces:
  - '**'
  additional_service_yaml: {}
  custom_dashboards:
    excludes:
    - ""
    includes:
    - '*'
external_services:
  istio:
    component_status:
      enabled: false
    root_namespace: d8-istio
    url_service_version: http://istiod.d8-istio:15014/version
    config_map_name: istio-v1x16
  prometheus:
    url: https://trickster.d8-monitoring/trickster/main
    auth:
      insecure_skip_verify: true
      type: bearer
      use_kiali_token: true
  grafana:
    url: http://prometheus.example.com
  custom_dashboards:
    enabled: true
identity: {}
istio_namespace: d8-istio
api:
  namespaces:
    exclude:
    - ^kube-system$
    - ^upmeter-probe-namespace-.+$
server:
  metrics_enabled: true
  metrics_port: 9090
  port: 20001
  web_root: /kiali

```

**Comment by nrfox** [comment]
Looking at the offending line(s) from the stacktrace: https://github.com/kiali/kiali/blob/885b0400273e44df529f3b2dc5f6c5d576cfa979/business/mesh.go#L596-L597

My guess is that `.Memory()` and `.Cpu()` both return nil when resource requirements are not set for istiod which causes a panic when `.AsApproximateFloat64()` is called on a nil pointer. It's valid for resource requirements to not be set on the deployment so Kiali should check for this.

**Comment by jshaughn** [comment]
@nrfox , +1, I had identified those lines up in a previous comment in January, not sure why they have not been protected.  Let's get that done for v1.66.

**Comment by leandroberetta** [comment]
> @nrfox , +1, I had identified those lines up in a previous comment in January, not sure why they have not been protected. Let's get that done for v1.66.

@jshaughn @nrfox I was waiting for some responses in this issue to start doing something, because some of the fields you indicated as optionals are always filled when Kiali starts (from the configmap or defaults in the code), so I wanted to check some config from the user. 

The offending line that was recently shared is a new one, and clearly it is something to proctect.

I apology myself for the delay @fl64 and thanks @nrfox for taking care of this during my holidays.

**Comment by nrfox** [comment]
> @jshaughn @nrfox I was waiting for some responses in this issue to start doing something, because some of the fields you indicated as optionals are always filled when Kiali starts (from the configmap or defaults in the code), so I wanted to check some config from the user.

For what it's worth, I wasn't able to reproduce the error either wrt limits being nil but it's worth adding checks for that anyways. I did not try deploying with helm though like what was reported if that makes a difference somehow.

**Comment by nrfox** [comment]
This should be fixed in the next release, `v1.66`, scheduled for Friday.
